api/opencl_runtime_layer.asciidoc

```diff
@@ -11434,10 +11395,10 @@ Otherwise, it returns one of the following errors:
     _num_events_in_wait_list_ is 0, or if event objects in _event_wait_list_
     are not valid events.
   * {CL_INVALID_OPERATION} if SVM pointers are passed as arguments to a kernel
-    and the device does not support SVM or if system pointers are passed as
-    arguments to a kernel and/or stored inside SVM allocations passed as
-    kernel arguments and the device does not support fine grain system SVM
-    allocations.
+    and the device does not support SVM, or if system pointers are passed as
+    arguments to a kernel and the device does not support fine-grain system SVM,
+    or if {CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM} is {CL_TRUE} and the device
+    does not support fine-grain system SVM.
```

Comment:
I'm not sure if the `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` part is really needed here. An client could use a kernel on multiple devices, some not supporting system SVM. Should we now require applications to flip `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` each time they enqueue the kernel on a different device?

It kinda feels more like an inconvenience, because for the runtime it's more or less meaningless as well. If an application uses a system SVM pointer on a device not supporting it indirectly, having `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` set to `CL_TRUE` or `CL_FALSE` doesn't really change anything about the abnormal behavior the client will see.

There is another aspect to this, by default `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` is `CL_TRUE`/`CL_FALSE` depending on the device. An application _might_ set it to `CL_FALSE` for devices supporting system SVM to be nice to the runtime, but maybe they want to revert that, so now they'll need to toggle, which earlier they probably wouldn't had to.

I think it would probably make more sense to simply state, that `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` is assumed to be always `CL_FALSE` on devices the kernel gets enqueued not supporting system SVM or that this flag only has an impact on kernels for devices supporting system SVM.

Though to be fair, I doubt there is any applications doing that, but so far it just sounds like an inconvenience gets added for no real benefit.

Reply:
Full disclosure: I've never liked `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM`.  Having the default state dependent on the capabilities of the device always seemed especially problematic.

This is an interesting idea:

> I think it would probably make more sense to simply state, that `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` is assumed to be always `CL_FALSE` on devices the kernel gets enqueued not supporting system SVM or that this flag only has an impact on kernels for devices supporting system SVM.

I think this could solve the default state problem as well, since then the default could be `CL_TRUE` if **any** device supports system SVM, and it'd just be ignored for devices that do not support system SVM.

If we're OK with this, I think I'd just need to make a few small edits to remove this error condition and to tidy up the `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` description.

Reply:
I think I've done this now, please take a look.  In short:

1. To use system SVM allocations indirectly within a kernel, the device must support system SVM, and `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` must be `CL_TRUE`.  If either of these conditions is false, accessing a system allocation within a kernel is undefined behavior.
    * Note specifically: There is no requirement to return an error if either condition is false.  Because behavior is undefined, returning an error is allowed, but not required.
3. The default value for `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` is `CL_TRUE`, no change here.
4. Trying to set `CL_KERNEL_EXEC_INFO_SVM_FINE_GRAIN_SYSTEM` to `CL_TRUE` if none of the devices in _context_ support system SVM is an error, no change here.